lua/neo-tree/sources/filesystem/lib/shared_clipboard.lua

```diff
@@ -0,0 +1,79 @@
+local fs_watch = require("neo-tree.sources.filesystem.lib.fs_watch")
+local manager = require("neo-tree.sources.manager")
+local events = require("neo-tree.events")
+local renderer = require("neo-tree.ui.renderer")
+local log = require("neo-tree.log")
+
+local M = {}
+
+local clipboard_state_dir_path = vim.fn.stdpath("state") .. "/neo-tree/"
+local clipboard_file_path = clipboard_state_dir_path .. "filesystem-clipboard.json"
+local clipboard_file_change_triggered_by_cur_neovim_instance = false
+
+M.save_clipboard = function(clipboard)
+  local file = io.open(clipboard_file_path, "w+")
+  -- We want to erase data in the file if clipboard is nil instead writing null
+  if not clipboard or not file then
+    return
+  end
+
+  local is_success, data = pcall(vim.json.encode, clipboard)
+  if not is_success then
+    log.error("Failed to save clipboard. JSON serialization error")
+    return
+  end
+  file:write(data)
+  file:flush()
+  M._update_all_cilpboards(clipboard)
+  clipboard_file_change_triggered_by_cur_neovim_instance = true
+end
+
+M._load_clipboard = function()
+  local file = io.open(clipboard_file_path, "r")
+  if not file then
+    return nil
+  end
+  local content = file:read("*a")
+  local is_success, clipboard = pcall(vim.json.decode, content)
+  if not is_success then
+    return nil
+  end
+  return clipboard
+end
+
+M._update_all_cilpboards = function(clipboard)
+  manager._for_each_state("filesystem", function(state)
+    state.clipboard = clipboard
+    vim.schedule(function()
+      renderer.redraw(state)
+    end)
+  end)
+end
+
+M.init = function()
+  if vim.fn.isdirectory(clipboard_state_dir_path) == 0 then
+    vim.fn.mkdir(clipboard_state_dir_path)
+  end
+
+  events.subscribe({
+    event = events.STATE_CREATED,
+    handler = function(state)
+      if state.name ~= "filesystem" then
+        return
+      end
+      vim.schedule(function()
+        M._update_all_cilpboards(M._load_clipboard())
+      end)
+    end,
+  })
+
+  -- Using watch_folder because we haven't "watch_file" :)
+  fs_watch.watch_folder(clipboard_state_dir_path, function()
+    if not clipboard_file_change_triggered_by_cur_neovim_instance then
```

Comment:
This is a bad idea. You can't say for sure that a file event came from this instance just because you wrote a file recently. This is fertile ground for hard to recreate bugs.

Between this and the modified time based check I would rather use the first solution where you just process every write, even your own.

I think I have a better idea:

1. When you convert the clipboard to json, save the json string as `last_clipboard_saved`
2. When you read in a new changed file event, compare the contents of the changed file to the `last_clipboard_saved`
3. If they are equal, nil out the `last_clipboard_saved` and ignore the event.

Reply:
Actually I can't understand why we can't be sure about this. Can you give me an example, please?

1. The variable `clipboard_file_change_triggered_by_cur_neovim_instance` is set in the `save_clipboard` function after writing the clipboard to the file
2. After that an event is immediately triggered in response to the file change.

Why can't we be sure that the write occurred in the current instance? We are not performing any other actions between these two operations, are we?

Reply:
We do not have exclusive access to this folder. Another program could write a file in this folder for some reason. 

For that matter, neo-tree itself might write a file here. It doesn't do that now, but someone else may do so in the future and they likely won't know that some other section of the code is naively reading every file written without checking the contents or even the file name.

Finally, I don't think you should be so confident about the order of events when you are dealing with asynchronous actions.